cloud.foundry.cli/src/main/java/cloud/foundry/cli/services/DumpController.java

```diff
@@ -0,0 +1,86 @@
+ * class that contains common behaviour of all the subcommands.
+ */
+@Command(name = "dump", header = "%n@|green Resolve " + RefResolver.REF_KEY + "-occurrences in a configuration file" +
+        " and print the result.|@",
+        mixinStandardHelpOptions = true, subcommands = {
+        DumpController.DumpAllCommand.class,
+        DumpController.DumpServicesCommand.class,
+        DumpController.DumpApplicationsCommand.class,
+        DumpController.DumpSpaceDevelopersCommand.class
+})
+public class DumpController implements Callable<Integer> {
+
+    // FIXME this field is actually not necessary, but needed due to the argument extension mechanism in the class
+    // CfArgumentsCreator... its contents are simply ignored by the dump command
+    @Mixin
+    private static LoginCommandOptions loginOptions;
+
+    @Override
+    public Integer call() {
+        usage(this, System.out);
+        return 0;
+    }
+
+    @Command(name = "all", description = "Resolve " + RefResolver.REF_KEY + "-occurrences in an entire " +
+            "configuration file and print the result.")
+    static class DumpAllCommand extends DumpCommandBase<ConfigBean> {
+
+        /**
+         * Constructor that specifies the bean type and the command class for {@link DumpCommandBase}
+         */
+        protected DumpAllCommand() {
+            super(ConfigBean.class, DumpAllCommand.class);
+        }
+    }
+
+    @Command(name = "services", description = "Resolve " + RefResolver.REF_KEY + "-occurrences in a service " +
+            "configuration file and print the result.")
+    static class DumpServicesCommand extends DumpCommandBase<SpecBean> {
+
+        /**
+         * Constructor that specifies the bean type and the command class for {@link DumpCommandBase}
+         */
+        protected DumpServicesCommand() {
+            super(SpecBean.class, DumpServicesCommand.class);
+        }
+    }
+
+    @Command(name = "applications", description = "Resolve " + RefResolver.REF_KEY + "-occurrences in an application " +
+            "configuration file and print the result.")
+    static class DumpApplicationsCommand extends DumpCommandBase<SpecBean> {
+
+        /**
+         * Constructor that specifies the bean type and the command class for {@link DumpCommandBase}
+         */
+        protected DumpApplicationsCommand() {
+            super(SpecBean.class, DumpApplicationsCommand.class);
+        }
+    }
+
+    @Command(name = "space-developers", description = "Resolve " + RefResolver.REF_KEY + "-occurrences in a " +
+            "space-developers configuration file and print the result.")
+    static class DumpSpaceDevelopersCommand extends DumpCommandBase<SpecBean> {
+
+        /**
+         * Constructor that specifies the bean type and the command class for {@link DumpCommandBase}
+         */
+        protected DumpSpaceDevelopersCommand() {
+            super(SpecBean.class, DumpSpaceDevelopersCommand.class);
```

Comment:
Not a big fan of having all these extra classes. In the other controllers, they're created inside here. The static wrapper classes are also defined here.

Reply:
We need an extra class for each subcommand, there is no way to avoid that. I reduced duplicate code as much as possible by moving the actual dump-implementation to the DumpCommandBase class. What do you mean by the static wrapper classes?

Reply:
> What do you mean by the static wrapper classes?

`static class DumpSpaceDevelopersCommand extends DumpCommandBase<SpecBean>`

> We need an extra class for each subcommand, there is no way to avoid that.

I think it might have been possible to move all the clases as static private ones into this class. Not pretty but that'll make it self-contained.

I think your way is okay, too. I'd suggest to move your dump-specific classes into a package, though.

Reply:
Every command already has its static class in the DumpController class (we could add the private modifier for sure). The difference is, that these derive from DumpCommandBase to inherit the dump functionality.